packages/ant-component-mapper/src/files/field-array.js

```diff
@@ -1,3 +1,209 @@
+  remove: 'REMOVE'
+};
+
+const initialState = {
+  index: 0,
+  history: []
+};
+
+export const reducer = (state, { type, action }) => {
+  switch (type) {
+    case 'redo':
+      return {
+        ...state,
+        index: state.index + 1
+      };
+    case 'action':
+      return {
+        index: state.index + 1,
+        history: [...state.history.slice(0, state.index), action]
+      };
+    case 'undo':
+      return {
+        ...state,
+        index: state.index - 1
+      };
+    case 'resetHistory':
+      return {
+        ...state,
+        history: state.history.slice(0, state.index)
+      };
+    default:
+      return state;
+  }
+};
+
+const DynamicArray = ({ ...props }) => {
+  const {
+    arrayValidator,
+    label,
+    description,
+    fields: formFields,
+    defaultItem,
+    meta,
+    minItems,
+    maxItems,
+    noItemsMessage,
+    FormFieldGridProps,
+    FormControlProps,
+    buttonLabels,
+    validateOnMount,
+    isRequired,
+    helperText,
+    FormItemProps,
+    ...rest
+  } = useFieldApi(props);
+  const [state, dispatch] = useReducer(reducer, initialState);
+
+  const combinedButtonLabels = {
+    ...defaultButtonLabels,
+    ...buttonLabels
+  };
+
+  const { dirty, submitFailed, error } = meta;
+  const isError = (dirty || submitFailed) && error && typeof error === 'string';
+  return (
+    <AntForm
+      meta={{ ...meta, error: typeof error === 'object' ? error.name : error }}
+      validateOnMount={validateOnMount}
+      helperText={helperText}
+      FormItemProps={FormItemProps}
+      isRequired={isRequired}
+    >
+      <FieldArray key={rest.input.name} name={rest.input.name} validate={arrayValidator}>
+        {({ fields: { map, value = [], push, remove } }) => {
+          const pushWrapper = () => {
+            dispatch({ type: 'resetHistory' });
+            push(defaultItem);
+          };
+
+          const removeWrapper = (index) => {
+            dispatch({ type: 'action', action: { action: 'remove', value: value[index] } });
+            remove(index);
+          };
+
+          const undo = () => {
+            push(state.history[state.index - 1].value);
+            dispatch({ type: 'undo' });
+          };
+
+          const redo = () => {
+            remove(value.length - 1);
+            dispatch({ type: 'redo' });
+          };
+
+          return (
+            <Row gutter={[0, 16]}>
+              <Col span={24}>
+                <Row justify="space-between">
+                  <Col>{label && <Typography.Title level={4}>{label}</Typography.Title>}</Col>
+                  <Col>
+                    <Space>
+                      <Button type="default" onClick={undo} disabled={state.index === 0} icon={<UndoOutlined />} />
+                      <Button type="default" onClick={redo} disabled={state.index === state.history.length} icon={<RedoOutlined />} />
+                      <Button type="primary" onClick={pushWrapper} disabled={value.length >= maxItems}>
```

Comment:
I think that these buttons could be customizable

Reply:
the same applies for remove button

Reply:
Right, I've completely neglected the customization. I will add it to all of the components.